src/main/java/net/openhft/chronicle/wire/internal/InternalAutoTailers.java

```diff
@@ -0,0 +1,100 @@
+import net.openhft.chronicle.threads.Pauser;
+import net.openhft.chronicle.wire.ExcerptListener;
+import net.openhft.chronicle.wire.MarshallableIn;
+import net.openhft.chronicle.wire.internal.reduction.ReductionUtil;
+import net.openhft.chronicle.wire.AutoTailers;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.function.Supplier;
+
+import static net.openhft.chronicle.core.util.ObjectUtils.requireNonNull;
+
+public final class InternalAutoTailers {
+
+    private InternalAutoTailers() {
+    }
+
+    public static final class RunnablePoller extends AbstractPoller implements AutoTailers.CloseableRunnable {
+
+        private final Pauser pauser;
+
+        public RunnablePoller(@NotNull final Supplier<? extends MarshallableIn> tailerSupplier,
+                              @NotNull final ExcerptListener excerptListener,
+                              @NotNull final Supplier<Pauser> pauserSupplier) {
+            super(tailerSupplier, excerptListener);
+            requireNonNull(pauserSupplier);
+            this.pauser = requireNonNull(pauserSupplier.get());
+        }
+
+        @Override
+        public void run() {
+            try {
+                while (running) {
+                    if (ReductionUtil.accept(tailer, excerptListener) != -1) {
+                        pauser.pause();
+                    }
+                }
+            } finally {
+                closer.run();
+            }
+        }
+    }
+
+    public static final class EventHandlerPoller extends AbstractPoller implements AutoTailers.CloseableEventHandler {
+
+        public EventHandlerPoller(@NotNull final Supplier<? extends MarshallableIn> tailerSupplier,
+                                  @NotNull final ExcerptListener excerptListener) {
+            super(tailerSupplier, excerptListener);
+        }
+
+        @Override
+        public boolean action() throws InvalidEventHandlerException {
+            if (!running) {
+                closer.run();
+                throw InvalidEventHandlerException.reusable();
+            }
+            return ReductionUtil.accept(tailer, excerptListener) != -1;
+        }
+    }
+
+    private abstract static class AbstractPoller implements AutoCloseable {
+
+        protected final ExcerptListener excerptListener;
+        protected final MarshallableIn tailer;
+        protected final Runnable closer;
+        protected volatile boolean running = true;
+
+        protected AbstractPoller(@NotNull final Supplier<? extends MarshallableIn> tailerSupplier,
+                                 @NotNull final ExcerptListener excerptListener) {
+            requireNonNull(tailerSupplier);
+            this.excerptListener = requireNonNull(excerptListener);
+            this.tailer = requireNonNull(tailerSupplier.get());
+            this.closer = closer(tailer);
+        }
+
+        @Override
+        public final void close() {
+            running = false;
```

Comment:
should this not call closer.run() on the current thread - as there look to be no guarantees that the closer is called? it relies on action() being called

Reply:
I think the Tailer can ony be used by one thread. So, the event handler needs to close it.